### PR TITLE
[bug] Update broadcast + reduce decision ModelCheckpoint]

### LINKED ISSUE #6343: trainer.training_type_plugin.broadcast doesn't seem to work properly
### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR update: 
* broadcast function to rely on `broadcast_object_list` from PyTorch.
* add `reduce_boolean_decision` for ModelCheckpoint | EarlyStopping using `all_gather`. Missing for Horovod.

Fixes #6343

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/distributed/dist.py [hunk 1]
```diff
@@ -31,32 +23,13 @@ def __init__(self, rank=None, device=None):
...
-        self._broadcast(data_tensor, src=0, group=group)
-        buffer = io.BytesIO(data_tensor.cpu().numpy())
-        obj = torch.load(buffer)
-        return obj
+    def broadcast(self, obj: Any, group=_group.WORLD):
+        # always wrap into a list so list can be brodcasted.
+        obj = [obj]
+
+        if self.rank != 0:
+            obj = [None for _ in range(len(obj))]
```
Comment (Borda) [inline pytorch_lightning/distributed/dist.py:31]:
```suggestion
            obj = [None] * len(obj)
```

File: pytorch_lightning/distributed/dist.py [hunk 2]
```diff
@@ -31,32 +23,13 @@ def __init__(self, rank=None, device=None):
...
-        self._broadcast(data_tensor, src=0, group=group)
-
-    def _receive(self, group=WORLD):
-        length_tensor = torch.tensor([0]).long().to(self.device)
-        self._broadcast(length_tensor, src=0, group=group)
-        data_tensor = torch.empty([length_tensor.item()], dtype=torch.uint8).to(self.device)
-        self._broadcast(data_tensor, src=0, group=group)
-        buffer = io.BytesIO(data_tensor.cpu().numpy())
-        obj = torch.load(buffer)
-        return obj
```
Comment (Borda) [inline pytorch_lightning/distributed/dist.py:62]: quite surprised that none of these is not needed anymore...

File: pytorch_lightning/utilities/distributed.py [hunk 3]
```diff
@@ -33,6 +37,87 @@ class group:
...
+    broadcast(object_tensor, src=src, group=group)
+
+    # Deserialize objects using their stored sizes.
+    offset = 0
+    if my_rank != src:
+        for i, obj_size in enumerate(object_sizes_tensor):
+            obj_view = object_tensor[offset:offset + obj_size]
+            obj_view = obj_view.type(torch.ByteTensor)  # type: ignore[call-overload]
+            offset += obj_size
+            object_list[i] = _tensor_to_object(obj_view, obj_size)
```
Comment (Borda) [inline pytorch_lightning/utilities/distributed.py:112]: as it is copy-paste from PT maybe move it to a special file `torch_distributed` when we drop out copywriter...?

**Comment by carmocca** [comment]
Is this a duplicate of https://github.com/PyTorchLightning/pytorch-lightning/pull/6345? They both have the same changes. What's the relation?

If one depends on the other, you should update one with `[Blocked by #...]`

File: pytorch_lightning/distributed/dist.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/distributed/dist.py:31]: Done !

File: pytorch_lightning/distributed/dist.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/distributed/dist.py:62]: It is replaced by PyTorch broadcast function which should be more robust. Somehow, torch.save didn't save the obj within the buffer, resulting in wrong broadcast.

**Comment by tchaton** [comment]
> Is this a duplicate of #6345? They both have the same changes. What's the relation?
> 
> If one depends on the other, you should update one with `[Blocked by #...]`

I opened 3 PRs from it resolving 3 different things and closed the previous one.

Best,
T.C

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -528,17 +533,9 @@ def _save_top_k_checkpoint(self, trainer, monitor_candidates: Dict[str, Any]):
...
-        # TODO: Move this logic to logger_connector. This also needs to be fixed for any
-        # other monitor logged value which aren't produced from a Metric.
-        if self.monitor == "val_loss":
-            current = trainer.training_type_plugin.reduce(current, reduce_op="mean")
-
-        if self.check_monitor_top_k(current):
+        if self.check_monitor_top_k(trainer, current):
             self._update_best_and_save(current, epoch, step, trainer, monitor_candidates)
-        elif self.verbose:
+        elif self.monitor is not None and self.verbose:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:538]: This seems to come from a bad merge. There is an early exit at the top of this fn if `self.monitor is None`

```suggestion
        elif self.verbose:
```

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 5]
```diff
@@ -11,15 +11,20 @@
...
 
 import torch
 
+import pytorch_lightning as pl
 from pytorch_lightning.core.step_result import Result
 from pytorch_lightning.trainer.states import TrainerState
 from pytorch_lightning.utilities import DistributedType, LightningEnum
 
+log = logging.getLogger(__name__)
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:27]: This doesn't seem to be used in this file

File: tests/special_tests.sh [hunk 6]
```diff
@@ -19,6 +19,7 @@ python ${DEFAULTS} tests/trainer/optimization/test_manual_optimization.py::test_
 python ${DEFAULTS} tests/models/test_sync_batchnorm.py::test_sync_batchnorm_ddp
 python ${DEFAULTS} tests/plugins/test_deepspeed_plugin.py::test_deepspeed_multigpu
 python ${DEFAULTS} tests/plugins/test_rpc_plugin.py::test_rpc_function_calls_ddp
+# todo (tchaton) tests are hanging in CI.
```
Comment (carmocca) [inline tests/special_tests.sh:22]: A GH issue might be better suited for this TODO

File: pytorch_lightning/plugins/training_type/parallel.py [hunk 7]
```diff
@@ -112,13 +115,3 @@ def block_backward_sync(self):
...
-
-    def broadcast(self, obj: object, src: int) -> object:
-        buffer = io.BytesIO()
-        torch.save(obj, buffer)
-        data = bytearray(buffer.getbuffer())
-        data_tensor = torch.tensor(data).to(self.root_device, dtype=torch.float)
-        data = all_gather_ddp_if_available(data_tensor)
-        buffer = io.BytesIO(data.cpu().byte().numpy())
-        obj = torch.load(buffer)
-        return obj
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/parallel.py:124]: Is this not necessary/unused now?

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:27]: Removed !

File: tests/special_tests.sh (hunk 6 above)
Comment (tchaton) [reply tests/special_tests.sh:22]: Removed !

File: pytorch_lightning/plugins/training_type/parallel.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/parallel.py:124]: All the parallel plugin overrides this function and therefore should not rely on it.

**Review by carmocca — APPROVED** [review]

**Comment by carmocca** [comment]
GPU tests are failing

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 8]
```diff
@@ -346,7 +346,12 @@ def check_monitor_top_k(self, current: torch.Tensor) -> bool:
             current = torch.tensor(current)
 
         monitor_op = {"min": torch.lt, "max": torch.gt}[self.mode]
-        return monitor_op(current, self.best_k_models[self.kth_best_model_path]).item()
+        should_update_best_and_save = monitor_op(current, self.best_k_models[self.kth_best_model_path])
+
+        # If using multiple devices, make sure all processes are unimanious on the decision.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:351]:
```suggestion
        # If using multiple devices, make sure all processes are unanimous on the decision.
```

File: pytorch_lightning/utilities/torch_distributed.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+from pytorch_lightning.utilities.imports import _TORCH_GREATER_EQUAL_1_7
+
+log = logging.getLogger(__name__)
+
+if torch.distributed.is_available():
+    from torch.distributed import Backend, broadcast, get_backend, get_rank, GroupMember
+
+# The code underneath is taken from PyTorch ``torch/distributed/distributed_c10d.py``
+# and enable broadcasting for PyTorch 1.6 and lower.
```
Comment (awaelchli) [inline pytorch_lightning/utilities/torch_distributed.py:14]: maybe this file should go to the "overrides" directory. 
could be good to have all customized pytorch code in one place.
not sure, let me know what you think.

File: tests/checkpointing/test_model_checkpoint.py [hunk 10]
```diff
@@ -366,8 +366,7 @@ def on_train_end(self, trainer, pl_module):
         assert self.best_model_score
         assert self.on_save_checkpoint_count == self.expected_count
         if trainer.is_global_zero:
-            # twice the calls expected because ddp broadcast also uses torch.save
-            assert torch.save.call_count == self.expected_count * 2
+            assert torch.save.call_count == self.expected_count
```
Comment (awaelchli) [inline tests/checkpointing/test_model_checkpoint.py:369]: great!

File: tests/checkpointing/test_checkpoint_callback_frequency.py [hunk 11]
```diff
@@ -102,3 +103,42 @@ def training_step(self, batch, batch_idx):
 
     # make sure types are correct
     assert save_mock.call_count == expected
+
+
+@mock.patch('torch.save')
+@RunIf(special=True, min_gpus=2)
+@pytest.mark.parametrize(['k', 'epochs', 'val_check_interval', 'expected'], [(1, 1, 1.0, 1), (2, 2, 0.3, 5)])
+def test_top_k_ddp(save_mock, tmpdir, k, epochs, val_check_interval, expected):
```
Comment (awaelchli) [inline tests/checkpointing/test_checkpoint_callback_frequency.py:111]: something is wrong with this test, shouldn't it fail with an error saying that the monitor key is not found?
`self.log uses "my_loss" but monitor="my_loss_step"`

File: pytorch_lightning/plugins/training_type/single_device.py [hunk 12]
```diff
@@ -23,6 +23,8 @@ class SingleDevicePlugin(TrainingTypePlugin):
     def __init__(self, device: torch.device):
         super().__init__()
         self.device: torch.device = device
+        self.world_size = 1
+        self.local_rank = 0
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/single_device.py:27]: is this required? where is it used?

File: tests/checkpointing/test_checkpoint_callback_frequency.py (hunk 11 above)
Comment (awaelchli) [reply tests/checkpointing/test_checkpoint_callback_frequency.py:111]: nvm! 
it logs both on step and epoch, therefore a _step postfix is added

File: pytorch_lightning/utilities/torch_distributed.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/utilities/torch_distributed.py:14]: Makes sense. Let me do that :)

File: pytorch_lightning/plugins/training_type/single_device.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/single_device.py:27]: Not in this PR, but the one to skip warning. Let me remove it.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 13]
```diff
@@ -33,7 +33,9 @@ class TrainingTypePlugin(Plugin, ABC):
     def __init__(self) -> None:
         self._model = None
         self._results = None
-        self.global_rank = 0
+        self.global_rank = None
+        self.local_rank = None
+        self.world_size = None
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:38]:
```suggestion
        self.global_rank = 0
        self.local_rank = 0
        self.world_size = 1
```
Can we keep it as before and not set None?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:38]: Hey @awaelchli. I actually think there is a bug behind this default, which could explain memory on gpu 0 when training on [1, 2] for example.
I will revert this back in the PR, but worth to investigate if everything is fine.
